src/plugins/ipt/ipt.cpp

```diff
@@ -0,0 +1,469 @@
+    int fail = 0;
+
+    uint8_t* buf = plugin->vcpu[vcpu_id].buf;
+    uint64_t size = plugin->vcpu[vcpu_id].size;
+    FILE* fd = plugin->vcpu[vcpu_id].fd;
+
+    if (!ret)
+    {
+        PRINT_DEBUG("annotate_ipt() failed to get ipt offset for vcpu %d\n", vcpu_id);
+        return 0;
+    }
+
+    PRINT_DEBUG("annotate_ipt() vCPU: %d IPT_CUR: %llx IPT_LAST: %llx\n", vcpu_id, (unsigned long long)offset, (unsigned long long)last_offset);
+
+    if (offset > last_offset)
+    {
+        fwrite(buf + last_offset, offset - last_offset, 1, fd);
+    }
+    else if (offset < last_offset)
+    {
+        fwrite(buf + last_offset, size - last_offset, 1, fd);
+        fwrite(buf, offset, 1, fd);
+    }
+    else
+    {
+        PRINT_DEBUG("annotate_ipt() called but no new IPT data is present\n");
+        fail = 1;
+    }
+
+    uint8_t ptwrite_packet[10] = {0x02, 0x32,};
+    uint32_t x1 = PTW_ERROR_EMPTY;
+    uint32_t x2 = 0;
+
+    for (size_t i = 0; i < num_payloads && !fail; i++)
+    {
+        x1 = payloads[(i * 2) + 1];
+        x2 = payloads[i * 2];
+
+        memcpy(&ptwrite_packet[2], &x1, sizeof(uint32_t));
+        memcpy(&ptwrite_packet[6], &x2, sizeof(uint32_t));
+        fwrite(ptwrite_packet, 10, 1, fd);
+    }
+
+    if (fail)
+    {
+        x1 = PTW_ERROR_EMPTY;
+        x2 = 0;
+
+        memcpy(&ptwrite_packet[2], &x1, sizeof(uint32_t));
+        memcpy(&ptwrite_packet[6], &x2, sizeof(uint32_t));
+        fwrite(ptwrite_packet, 10, 1, fd);
+    }
+
+    return 1;
+}
+
+struct exec_fault_data
+{
+    ipt* plugin;
+    addr_t rip;
+};
+
+struct access_fault_result_t: public call_result_t
+{
+    access_fault_result_t() : call_result_t(), fault_va() {}
+
+    addr_t fault_va;
+};
+
+static event_response_t execute_faulted_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    struct exec_fault_data* ef_data = (struct exec_fault_data*)info->trap->data;
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+        .addr = ef_data->rip
+    };
+
+    size_t bytes_read = 0;
+    uint8_t pagebuf[4096] = {0,};
+
+    vmi_read(vmi, &ctx, 4096, pagebuf, &bytes_read);
+
+    std::stringstream ss;
+    ss << ef_data->plugin->ipt_dir;
+    ss << "/frames/frame_";
+    ss << std::setfill('0') << std::setw(5) << ef_data->plugin->frame_num;
+
+    std::string frame_fn = ss.str();
+
+    FILE* fp = fopen(frame_fn.c_str(), "wb");
+
+    if (!fp)
+    {
+        PRINT_DEBUG("[IPT] Frame save path not accessible: %s\n", frame_fn.c_str());
+        drakvuf_release_vmi(drakvuf);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    fwrite(pagebuf, 1, bytes_read, fp);
+    fclose(fp);
+
+    uint64_t tsc = __rdtsc();
+
+    mmvad_info_t mmvad;
+    unicode_string_t* dll_name = nullptr;
+    char* dll_name_str = nullptr;
+    char missing_dll_name[] = "(null)";
+
+    addr_t base_va = 0;
+    addr_t end_va = 0;
```

Comment:
boy, this sure looks like it should be a structure or smth

Reply:
this is only printed and nothing else more is done with these two values, so is it worth to wrap with a struct here?